File: php/108-convert-sorted-array-to-binary-search-tree.php
<?php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution 
{
    /**
     * @param Integer[] $nums
     * @return TreeNode
     */
    public function sortedArrayToBST($nums) 
	{
        return $this->toTree(0, count($nums)-1, $nums);
    }
    public function toTree($left, $right, &$nums) 
	{
        if ($left > $right) return null;
        if ($left == $right) return new TreeNode($nums[$left]);
        
        $middle = ($left+$right)>>1;
        $node = new TreeNode($nums[$middle]);
        $node->left = $this->toTree($left, $middle-1, $nums);
        $node->right = $this->toTree($middle+1, $right, $nums);
        
        return $node;
    }
}

File: go/108-convert-sorted-array-to-binary-search-tree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	return toTree(0, len(nums)-1, &nums)
}
func toTree(left, right int, nums *[]int) *TreeNode {
	if left > right {
		return nil
	}
	if left == right {
		return &TreeNode{Val: (*nums)[left]}
	}
	middle := (left + right) / 2

	node := &TreeNode{Val: (*nums)[middle]}
	node.Left = toTree(left, middle-1, nums)
	node.Right = toTree(middle+1, right, nums)

	return node
}
